parser_library/src/workspaces/macro_cache.cpp

```diff
@@ -0,0 +1,202 @@
+    std::array<context::id_index, 19> cached_instr { wn.COPY,
+        wn.ASPACE,
+        wn.GBLA,
+        wn.GBLB,
+        wn.GBLC,
+        wn.LCLA,
+        wn.LCLB,
+        wn.LCLC,
+        wn.SETA,
+        wn.SETB,
+        wn.SETC,
+        ctx.ids().add("AIF"),
+        ctx.ids().add("MEND"),
+        ctx.ids().add("MACRO"),
+        ctx.ids().add("MEXIT"),
+        ctx.ids().add("AIF"),
+        ctx.ids().add("AREAD"),
+        ctx.ids().add("ACTR"),
+        ctx.ids().add("AGO") };
+
+    std::vector<cached_opsyn_mnemo> result;
+
+    for (const auto& [from, opcode] : ctx.opcode_mnemo_storage())
+    {
+        // If there is an opsyn, that aliases an instruction to be CA instruction, add it to result
+        if (std::find(cached_instr.begin(), cached_instr.end(), opcode.opcode) != cached_instr.end())
+            result.push_back(
+                { from, opcode.opcode, std::holds_alternative<context::macro_def_ptr>(opcode.opcode_detail) });
+
+        if (std::find(cached_instr.begin(), cached_instr.end(), from) != cached_instr.end())
+            result.push_back(
+                { from, opcode.opcode, std::holds_alternative<context::macro_def_ptr>(opcode.opcode_detail) });
+    }
+
+    // Also macros with the same name as CA instructions may alias the instructions
+    for (const auto& [id, macro] : ctx.macros())
+    {
+        if (!macro)
+            continue;
+        auto opcode = ctx.get_operation_code(id);
+        // Macros for which there is an opsyn in effect are already captured
+        if (auto pval = std::get_if<context::macro_def_ptr>(&opcode.opcode_detail); !pval || *pval != macro)
+            continue;
+
+        // If there is an opsyn, that aliases an instruction to be CA instruction, add it to result
+        if (std::find(cached_instr.begin(), cached_instr.end(), macro->id) != cached_instr.end())
+            result.push_back({ id, id, true });
+    }
+
+    sort_opsyn_state(result);
+
+    return result;
+}
+
+macro_cache_key macro_cache_key::create_from_context(context::hlasm_context& hlasm_ctx, library_data data)
+{
+    return { hlasm_ctx.opencode_file_name(), data, get_opsyn_state(hlasm_ctx) };
+}
+
+void macro_cache_key::sort_opsyn_state(std::vector<cached_opsyn_mnemo>& opsyn_state)
+{
+    std::sort(opsyn_state.begin(), opsyn_state.end(), [](const cached_opsyn_mnemo& lhs, const cached_opsyn_mnemo& rhs) {
+        return std::tie(lhs.from_instr, lhs.to_instr, lhs.is_macro)
+            < std::tie(rhs.from_instr, rhs.to_instr, rhs.is_macro);
+    });
+}
+
+const analyzer* macro_cache::find_cached_analyzer(const macro_cache_key& key) const
+{
+    auto it = cache_.find(key);
+    if (it == cache_.end())
+        return nullptr;
+
+    const auto& cached_data = it->second;
+
+    for (const auto& [fname, cached_version] : cached_data.stamps)
+    {
+        auto file = file_mngr_->find(fname);
+        if (!file)
```

Comment:
could we keep like weak_ptr to the file instead of file_manager ref and the name?

Reply:
Probably not, since file_manager creates new shared pointers to processor_files in the rare cases when the files are accessed concurrently. We need to access the newest version here.